### ISSUE TITLE
YOLO Training for Object Detection

### ISSUE THREAD

**Issue by TianF0F** [TianF0F/skills-github-pages#1]
Authors: Yue Tang, Tianfang Fang

YOLO, which stands for You Only Look Once, is a state-of-the-art object detection model known for high speed and accuracy in detecting objects within images. Be different from traditional detection methods such as R-CNN using a two-stage pipeline: first stage for region proposal and second stage for classification, YOLO is designed to perform both localization (find bounding boxes) and classification (assigning class labels) in a single step through the network. 

YOLO’s single-stage detection is achieved by dividing the input image into a grid and making predictions for each grid cell. It outputs both the object class and the bounding box coordinates directly, making it a perfect choice for real-time object detection tasks in a wide variety of application of autonomous driving, robotics, surveillance and video analysis where fast inference is crucial. 

We used PASCAL VOC 2007 dataset for our training. It contains annotated images with 20 object categories, including people, animals, vehicles, and household items. The dataset includes approximately 10,000 images with object bounding boxes, class labels, and segmentation masks. We load the dataset and preprocesses the images by converting them into PyTorch tensors and resize them to 224x 224. The low resolution is beneficial for computational efficiency while a real object detection system usually operates on higher-resolution images. 

After the dataset loading, we set up a coordinate system based on the CNN activation map with a shape of 7x7. As the top-left corner represented by (0,0) and the bottom-right corner represented by (7,7), all spatial operations during model processing can be aligned. We visualize parts of our dataset before building the model and here are some sample images with ground-truth object boxes and category labels visualized:
<img width="884" alt="Screenshot 2024-10-20 at 3 37 48 PM" src="https://github.com/user-attachments/assets/492850f8-497a-4452-8e6f-067c6a354440">

For image feature extraction we used MobileNet v2. Anchors in YOLO are predefined bounding boxes of different sizes used to detect objects in various parts of the image. With the spatial sizes of anchors been defined and a grid generated that split the whole image, we can compute the positions of all anchor boxes for the entire image. 

All nine anchors be exactly centered: torch.Size([9, 1, 1, 4]) and all nine anchors cluttered: torch.Size([9, 7, 7, 4]) respectively:
<img width="561" alt="Screenshot 2024-10-20 at 3 39 05 PM" src="https://github.com/user-attachments/assets/4114021f-105f-4e85-9c17-aca9056a83b1">

During training, we compute the ground-truth transformation that would transform the anchor box into the ground-truth box. We then apply a regression loss that penalizes differences between the predicted transform and the ground-truth transform. We need to match the ground-truth boxes against the anchors to determine the classification labels for the anchors for which anchors should be classified as containing an object and which should be classified as background. 

The prediction network inputs the features from the backbone network, and outputs the classification scores and transformations for each anchor. In the part of loss function, the confidence score regression loss is for both activated and negative anchors while the bounding box regression loss and the object classification loss are for activated anchors only. 

We then implement functions of our detector and detection solver object to train the detector in loops. To make sure that everything is working as expected, we try to overfit the detector to a small subset of data and the inference for overfit small data is shown below:
<img width="870" alt="Screenshot 2024-10-20 at 3 39 30 PM" src="https://github.com/user-attachments/assets/fe865387-ef8a-4fe5-bbd9-bc7e91bc1bf7">

Inference for net training:
<img width="880" alt="Screenshot 2024-10-20 at 3 39 52 PM" src="https://github.com/user-attachments/assets/e1d48182-5f73-4b94-935c-5eac06bd954c">

From preparing the dataset and visualizing ground truth bounding boxes to generating anchors of various sizes and training the model over multiple epochs, we built an efficient object detection pipeline. Through careful tuning and visualization of the results, we ensured that the model not only learned effectively but also demonstrated strong performance in detecting objects in images, marking the success of our object detection training process.
